ntoskrnl/config/cmparse.c

```diff
@@ -989,40 +1127,711 @@ CmpHandleExitNode(IN OUT PHHIVE *Hive,
+ * registry key path name. The remaining name is updated
+ * by the function if a key pathname is found in cache.
+ *
+ * @param[in,out] OuterStackArray
+ * A pointer to an array that lives on the caller's stack.
+ * The expected size of the array is up to 32 elements,
+ * which is the imposed limit by CMP_HASH_STACK_LIMIT.
+ * This limit also corresponds to the maximum depth of
+ * subkey levels.
+ *
+ * @param[in,out] Kcb
+ * A pointer to a KCB, this KCB is changed if the key pathname
+ * is found in cache.
+ *
+ * @param[out] Hive
+ * A pointer to a hive, this hive is changed if the key pathname
+ * is found in cache.
+ *
+ * @param[out] Cell
+ * A pointer to a cell, this cell is changed if the key pathname
+ * is found in cache.
+ *
+ * @param[out] MatchRemainSubkeyLevel
+ * A pointer to match subkey levels returned by the function.
+ * If no match levels are found, this is 0.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if cache lookup has completed successfully.
+ * STATUS_OBJECT_NAME_NOT_FOUND is returned if the current KCB of
+ * the key pathname has been deleted. STATUS_RETRY is returned if
+ * at least the current KCB or its parent have been deleted
+ * and a cache lookup must be retried again.
+ *
+ * @remarks
+ * The function attempts to do a cache lookup with a shared lock
+ * on KCBs so that other threads can simultaneously get access
+ * to these KCBs. When the captured KCB is being deleted on us
+ * we have to retry a lookup with exclusive look so that no other
+ * threads will mess with the KCBs and perform appropriate actions
+ * if a KCB is deleted.
+ */
 NTSTATUS
 NTAPI
-CmpBuildHashStackAndLookupCache(IN PCM_KEY_BODY ParseObject,
-                                IN OUT PCM_KEY_CONTROL_BLOCK *Kcb,
-                                IN PUNICODE_STRING Current,
-                                OUT PHHIVE *Hive,
-                                OUT HCELL_INDEX *Cell,
-                                OUT PULONG TotalRemainingSubkeys,
-                                OUT PULONG MatchRemainSubkeyLevel,
-                                OUT PULONG TotalSubkeys,
-                                OUT PULONG OuterStackArray,
-                                OUT PULONG *LockedKcbs)
+CmpLookInCache(
+    _In_ PCM_HASH_CACHE_STACK HashCacheStack,
+    _In_ BOOLEAN LockKcbsExclusive,
+    _In_ ULONG TotalRemainingSubkeys,
+    _Inout_ PUNICODE_STRING RemainingName,
+    _Inout_ PULONG OuterStackArray,
+    _Inout_ PCM_KEY_CONTROL_BLOCK *Kcb,
+    _Out_ PHHIVE *Hive,
+    _Out_ HCELL_INDEX *Cell,
+    _Out_ PULONG MatchRemainSubkeyLevel)
 {
-    /* We don't lock anything for now */
-    *LockedKcbs = NULL;
+    LONG RemainingSubkeys;
+    ULONG TotalLevels;
+    BOOLEAN SubkeysMatch;
+    PCM_KEY_CONTROL_BLOCK CurrentKcb, ParentKcb;
+    PCM_KEY_HASH HashEntry = NULL;
+    BOOLEAN KeyFoundInCache = FALSE;
+    PULONG LockedKcbs = NULL;
+
+    /* Reference the KCB */
+    (VOID)CmpReferenceKeyControlBlock(*Kcb);
```

Comment:
I'm not sure that ignoring the returned value is a good thing (especially since on return, the kcb is dereferenced).